### ISSUE TITLE
Class requires Angular annotation even when abstract.

### ISSUE THREAD

**Issue by zehavibarak** [angular/angular#37769]
Angular 10.0.0 build yields
> NG2007: Class is using Angular features but is not decorated
 while class is marked as abstract.

```
export abstract class B {}

@Injectable({...})
export class A extends B {}
```
CLI should be able to compile, as did earlier versions.

**Comment by JoostK** [comment]
This pattern [has been deprecated in v9](https://v9.angular.io/guide/deprecations#undecorated-base-classes-using-angular-features) and is no longer supported in v10.

There is a migration that typically migrates these cases automatically (see [documentation here](https://angular.io/guide/migration-undecorated-classes) and [also here](https://angular.io/guide/migration-injectable) for datails), however it may be that the inherited `@Injectable()` case is not handled.

Anyway, the resolution here is to add `@Injectable()` to the abstract base class.

**Comment by matthewjh** [comment]
@JoostK: are you sure it is needed? Note that the docs do _not_ say that `@Injectable` is needed in this case:

```As of version 9, it's deprecated to have an undecorated base class that:

uses Angular features
is extended by a directive or component
Angular lifecycle hooks or any of the following Angular field decorators are considered Angular features:

@Input()
@Output()
@HostBinding()
@HostListener()
@ViewChild() / @ViewChildren()
@ContentChild() / @ContentChildren()
```

For people who use abstract base classes for injectable services, none of these apply (not extended by a directive or component, not using any of those Angular decorators/features etc.)

**Comment by zehavibarak** [comment]
It did work in v9. In our project base classes are marked as **abstract**, and does not reference one of the annotations above. However, they do implement OnDestroy.

I see no reason for this to be non legit syntax.

**Comment by JoostK** [comment]
@matthewjh That sounds correct to me, yes. The trigger for the error in this case is the `ngOnDestroy` method, which is an Angular lifeycle hook which means that the class has Angular behavior; therefore requiring the `@Injectable()`. It could be that the documentation is not complete wrt this case.

@zehavibarak For the Angular runtime to execute the `ngOnDestroy` hook, it needs to be aware that it exists.

**Comment by zehavibarak** [comment]
Again, it did work in 9 and it is a good programming practice. Whatever change made in 10 will need to be meaningful for me to fold.

**Comment by JoostK** [comment]
@zehavibarak What practice are you referring to as "good programming practice"?

**Comment by nartc** [comment]
@JoostK Using `abstract` classes which are meant to be extended only. The following example worked in Angular 9

```typescript
export abstract class Destroyable implements OnDestroy {
  protected $destroyed = new Subject();

  ngOnDestroy() {
    this.$destroyed.next();
    this.$destroyed.complete();
  }
}

// SomeComponent
@Component({...})
export class SomeComponent extends Destroyable implements OnInit {
   ngOnInit() {
      someObservable.pipe(takeUntil(this.$destroyed)).subscribe(...)
   }
}
```
I think abstract classes should be excluded from the list of required annotations since it's never meant (or be able to) to be instantiated anyway.

**Comment by matthewjh** [comment]
@nartc I think what @JoostK is saying is that Angular _needs_ (or will in future need) the base class in your case to be annotated, so that the compiler knows to statically analyse it for Angular features like `ngOnDestroy`. Then, when the compiler sees your sub class extending `Destroyable`, it, having analysed the base class, knows that the aubclass will have an `ngOnDestroy` hook via `Destroyable`. This makes sense.

@JoostK just one thing I want to clarify, if we use "abstract class" like interfaces, and have a lot of classes that _implement_ (not extend) them, we won't need to add `@Injectable()` to the abstract classes, right?

**Comment by nartc** [comment]
@matthewjh isn't that what Ivy does? And also, I'm not against adding a selector-less decorator like `Directive` but the above snippet did work in Angular 9. So either this is a regression issue or Ivy contains a "bug as a feature". As @zehavibarak said, some explanation on why the above worked in 9 but stopped working in 10 would be nice. 

"worked in 9" as in sub class has `ngOnDestroy` hook ran just fine.

**Comment by JoostK** [comment]
@matthewjh Correct, your analysis is spot on.

> just one thing I want to clarify, if we use "abstract class" like interfaces, and have a lot of classes that implement (not extend) them, we won't need to add @Injectable() to the abstract classes, right?

Correct. It is fine to omit the decorator if the base class does not implement any Angular behavior, which an abstract base class used as interface + token won't have. Even abstract classes that are actually extended from other classes will only need a decorator if they use Angular features. There's two gotchas to this statement:

1. When a base class class has a constructor that has to be called by Angular, it needs either `@Injectable()` or `@Directive()` to capture those parameters as tokens for the runtime to use. This is a gotcha, as having a constructor won't require a decorator _iff_ all subclasses have their own constructor.
2. A class `B` that is part of an inheritance chain where its superclass `A` has Angular behavior and a subclass `C` has Angular behavior requires that class `B` also has a decorator, regardless of whether it has Angular behavior in itself. So "extending a class with Angular behavior" should also be considered as having Angular behavior. This case is probably not documented, I reckon.

@nartc These changes fall under the "[Undecorated base classes using Angular features](https://v9.angular.io/guide/deprecations#undecorated-base-classes-using-angular-features)" deprecations in v9. Support was removed in #36921. Note that it's absolutely fine to keep the `abstract` keyword.

> Or like you said, this might be needed in the future and adding annotation to abstract class is a mean to add more metadata for Angular Compiler.

The future is now! This is exactly what happened in v10.

**Comment by thw0rted** [comment]
I read through the whole thread as well as [the docs page](https://v9.angular.io/guide/migration-undecorated-classes) and I'm still not clear on when to use `@Injectable()` vs `@Directive()`.  Is the end result behavior any different?

I had an `abstract` base class that several Components inherited from.  The child classes were decorated `@Component(...)` and the base had no decorator.  As previous commenters pointed out, this seemed fine in v9 but broke in v10.  The docs I linked above don't give this specific example -- there's one where the base and child classes both have `@Component` but there's no such thing as a zero-argument `@Component()` decorator and it doesn't make sense to attach a `selector` to an abstract class (right?).  Given what I found, I decided to mark the base with an empty `@Directive()` decorator, which seems to work, but that feels wrong too -- it's a Component, not a Directive!

Is the real issue here that we "should" have an empty `@Component()` decorator?  Or am I over-thinking it?

**Comment by zehavibarak** [comment]
You are overthinking it. Works with empty @Directive()

**Comment by nartc** [comment]
@thw0rted `Component` is a `Directive` with `template`. Does your base have `template`? No? Then it is a `Directive`

**Comment by thw0rted** [comment]
Yeah, you're right, Directive makes more sense than Component because as a base class it's all functionality and no style or content / template.  But really it's not even a Directive, because isn't a Directive just functionality (???) that can attach to a host Element?  And of course an abstract "component base" can't attach to anything.  I think I'm talking myself into thinking that `Injectable` is more correct, or at least less wrong.  Maybe I just don't have a deep enough understanding of internals to grok this.

**Comment by Christopherjl** [comment]
In my opinion `Component`, `Directive` and `Injectable` does not make any sence on an `abstract class`. Clearly this is error is a false negative and adding `Directive` may be a workaround.

**Comment by john182** [comment]
I agree with @Christopherjl,  it makes no sense to write down an abstract class with "Component, Directive and Injectable"

**Comment by Lampent** [comment]
Any update on this?

**Comment by luizfbicalho** [comment]
If the base class is anstract and Generic, then ng build --prod with "enableIvy": false, to make a library, doesn't compile
then I have to remove the directive to make it compile, but if it's not production, still makes the NG2007 error.

**Comment by thw0rted** [comment]
What does `enableIvy: false` do now?  I thought Ivy was the only template engine available in v10+.

**Comment by luizfbicalho** [comment]
@thw0rted when I compile a library as a NPM package it can't be compiled with Ivy.

**Comment by ericsnap** [comment]
FWIW, I am running into issues related to this. I have no problem adding a `@Directive()` to my base class if it makes Angular happy. It is an ok compromise for me.

However, my use case has 2 specific conditions that may not be supported (or I need clarification on how to). A long time ago, we followed a pattern to implement custom control via `ControlValueAccessor`. I cannot find the original article we followed, but it has been working fine for a several Angular versions (6+) where we implement a _base class_ `ValueAccessorBase<T>` similar to:

```typescript
export abstract class ValueAccessorBase<T> implements ControlValueAccessor {
  @Input disabled: boolean;

  constructor(private writeValueCallback?: (value: T) => void) {};
  // the rest of the implementation of CVA
}
```
When we compile with Angular 10, we get the following error:
```
ERROR: /path/to/project/my-library-project/src/lib/value-accessor-base.ts:8:1: Error encountered in metadata generated for exported symbol 'ValueAccessorBase': 
 /path/to/project/my-library-project/src/lib/value-accessor-base.ts:17:46: Metadata collected contains an error that will be reported at runtime: Expression form not supported.
  {"__symbolic":"error","message":"Expression form not supported","line":16,"character":45}
An unhandled exception occurred: /path/to/project/my-library-project/src/lib/value-accessor-base.ts:8:1: Error encountered in metadata generated for exported symbol 'ValueAccessorBase': 
 /path/to/project/my-library-project/src/lib/value-accessor-base.ts:17:46: Metadata collected contains an error that will be reported at runtime: Expression form not supported.
  {"__symbolic":"error","message":"Expression form not supported","line":16,"character":45}
See "/private/var/folders/q9/9pm0lc5x3s5bw0nywxv0r9_c0000gr/T/ng-im3YyK/angular-errors.log" for further details.

```

The issue here seems to be the constructor parameter (a generic callback). I think I understand why, but I don't know how to solve this. One option could be to remove the `@Input()` on our `disabled` flag. We could still provide a protected variable or something to manage the `setDisabled()` CVA implementation. 

Any reason why that constructor callback is causing metadata errors? is it because of the generic part? I did try replacing it with `any` and it works, but obviously, that's not correct. Is there a decorator I can add to ignore the token injection?

**Comment by JoostK** [comment]
@ericsnap That error is reported by the View Engine compiler when compiling your library. It prepares a `metadata.json` file for an application build to use, but it notices that there is an error in this file that would prevent apps from compiling so it reports an error immediately during the library build. You should be able to set Angular compiler option `strictMetadataEmit: false` to disable this check: it will still result in a metadata file that would introduce a compilation error when consumed in an application, but that is only a problem if the app is still on View Engine. For apps using Ivy the metadata isn't used and I hope that you'd still be able to use this pattern.

**Comment by ericsnap** [comment]
@JoostK thanks for the super quick response. So, you are saying that we're not doing anything completely crazy, right? 
Since this is an internal library and we have some relative tight control over our consumers, _if_ we compile this library as-is using the Ivy engine _and_ make sure our projects use Ivy too, we should be good?

**Comment by JoostK** [comment]
> @JoostK thanks for the super quick response. So, you are saying that we're not doing anything completely crazy, right?

Nope, there is nothing crazy about that. But I do anticipate that you'll hit #37914, which would then require a workaround of setting `strictInjectionParameters: false`.

> Since this is an internal library and we have some relative tight control over our consumers, _if_ we compile this library as-is using the Ivy engine _and_ make sure our projects use Ivy too, we should be good?

Sort of, yes. However, it is not recommended to be publishing libraries directly using the Ivy compiler, as the generated code is not stable and can change between patch versions. I think you should be able to continue to compile libraries using View Engine as you do today, if you set `strictMetadataEmit: false`.

**Comment by thw0rted** [comment]
Can't you just use a `protected abstract writeValueCallback: (val: T) => void` as a member, instead of passing it as a constructor argument?  If the problem is that you need to use it in common setup code, you could move that logic into an Angular lifecycle event instead.

**Comment by ericsnap** [comment]
@thw0rted you are absolutely correct. That's what I wanted to do first. But it would require us/me to go through all our custom controls and manually update the constructor call to use a protected implementation instead. Considering what @JoostK mentioned above and #37914 I may put that in the backlog sooner rather than later.
The good news is that I don't see a lot of our controls with actual meaningful logic in that callback. Very few of our controls actually change the data in and out.

**Comment by distante** [comment]
> I read through the whole thread as well as [the docs page](https://v9.angular.io/guide/migration-undecorated-classes) and I'm still not clear on when to use `@Injectable()` vs `@Directive()`. Is the end result behavior any different?
> 
> I had an `abstract` base class that several Components inherited from. The child classes were decorated `@Component(...)` and the base had no decorator. As previous commenters pointed out, this seemed fine in v9 but broke in v10. The docs I linked above don't give this specific example -- there's one where the base and child classes both have `@Component` but there's no such thing as a zero-argument `@Component()` decorator and it doesn't make sense to attach a `selector` to an abstract class (right?). Given what I found, I decided to mark the base with an empty `@Directive()` decorator, which seems to work, but that feels wrong too -- it's a Component, not a Directive!
> 
> Is the real issue here that we "should" have an empty `@Component()` decorator? Or am I over-thinking it?

I have a similar doubt to this but regarding Services.

I have an `abstract` class that is used to create component level provided services. [The current migration guide](https://angular.io/guide/migration-undecorated-classes) does not have this use case.

```
abstract class BaseService implements OnDestroy {
    protected readonly subscriptions = new Subscription();

   ngOnDestroy() {
      this.subscriptions.unsubscribe();
  }
}


@Injectable()
class MyServiceImplementationA extends BaseService {
   // ...
} 

````

Should I decorate `BaseService`  with `Directive` ? It does not sound right.

**Comment by petebacondarwin** [comment]
You can apply "any" Angular decorator to ensure that the Angular compiler is aware of this class and so wires up the life-cycle hook correctly. In the case of services the recommendation would be to use `@Injectable()`.

Error: https://stackblitz.com/edit/angular-ivy-8rstse?file=src%2Fapp%2Fapp.module.ts
Fixes: https://stackblitz.com/edit/angular-ivy-yqqdce?file=src%2Fapp%2Fapp.module.ts


Perhaps someone could create an issue to document this case better?

**Comment by distante** [comment]
> You can apply "any" Angular decorator to ensure that the Angular compiler is aware of this class and so wires up the life-cycle hook correctly. In the case of services the recommendation would be to use `@Injectable()`.
> 
> Error: https://stackblitz.com/edit/angular-ivy-8rstse?file=src%2Fapp%2Fapp.module.ts
> Fixes: https://stackblitz.com/edit/angular-ivy-yqqdce?file=src%2Fapp%2Fapp.module.ts
> 
> Perhaps someone could create an issue to document this case better?

`@Injectable()` tells me that "this is injected" instead of "extended" though. But I suppose this is the best option right now. Maybe a `@AngularAware()` decorator for abstract classes would be clearer (Instead of using decorators that meant *just* one particular thing before).

**Comment by petebacondarwin** [comment]
I guess you could read `Injectable` as "can take part in the dependency injection system".

**Comment by kschaefe** [comment]
I have had to remove generic type parameters on my abstract parent components because Angular 11 will not compile correctly with them.  Then, I had to loosen types or cast in many locations throughout the codebase.

Ivy's inability to handle generic type parameters is a giant step backward in terms of code quality and safety.  How this is a P4, backlog item is mind-boggling.  Why even bother to TypeScript if you're not going to actually TypeScript?

**Comment by mlc-mlapis** [comment]
@kschaefe What do you mean concretely?

**Comment by JoostK** [comment]
@kschaefe  that doesn't sound right; generic type parameters shouldn't be causing issues. Could you please provide examples of code that you experienced your issues with, along with the compiler errors you got?

**Comment by hettiger** [comment]
If you need to provide some arbitrary value such as an API URL as a constructor parameter from the derived service class to the abstract super class I might have found a workaround:

```ts
@Injectable()
abstract class HttpService<Model> implements OnDestroy {
  constructor(
    @Inject('') private readonly apiUrl: string,
    private http: HttpClient,
  ) {}

  ngOnDestroy(): void {}
}

@Injectable({
  providedIn: 'root'
})
class ItemService extends HttpService<Item> {
  constructor(http: HttpClient) {
    super('https://…', http);
  }
}
```

The trick is using `@Inject('')` to silence the injection token error …

Since the class is abstract Angular will not inject anything here so this should be fine … (we're providing the value ourselves using the `super()` call)

I've tested this on a prod build and can confirm `ngOnDestroy()` is getting called and everything's working as intended.

To be honest: using `@Inject('')` feels super weird. It's a dirty workaround IMHO. However using `@Injectable()` on an abstract class ist not better in that regard. However, your mileage may vary.

Please let me know if the above workaround could cause any issues that I'm not aware of.

Maybe we can get some decorator to opt out from dependency injection or something along those lines in the future …? (Angular would still be getting the metadata that way …)

**Comment by Azbesciak** [comment]
I see that this issue is rather dead, so I will put something here.
I have an abstract class which has destructing logic
it is like this:
```ts
export abstract class Subscriber {
  private readonly subscriptions = new Subscriptions(); // some management container, not important

  ngOnDestroy(): void {
  /// heavy work.
  }
```

Of course, I get NG2007. I added `@Injectable`, no problem. Some time later (now) I need to generate lib.
And in general everything works, except cases when it does not, and I do not know why. But I have other error:
````
Uncaught (in promise): TypeError: Cannot set property ɵfac of class Subscriber {
  constructor() {
    this.subscriptions = new _imap_bundle_shared_flow_utils_subscriptions_...<omitted>...
} which has only a getter
````
Similar to https://stackoverflow.com/questions/60633963/angular-9-production-error-cannot-set-property-%C9%B5fac-of-abstract-class-myfilte, but I suppose that my case is rather 'easies' - I do not have directives, just base method which is named like angular one.

**Comment by slavede** [comment]
Try using @Directive above your abstract classes. Works fine for me. Still can be used as a parent of other injectables.

**Comment by Azbesciak** [comment]
@slavede thank you, works for me, although it is still neither clean nor intuitive...

**Comment by luizfbicalho** [comment]
> @slavede thank you, works for me, although it is still neither clean nor intuitive...

I agree, should not validate the abstract classes

**Comment by JeanMeche** [comment]
I'll close this as solved as this original issue is solved. The compiler doesn't complain anymore if there is a missing decorator on a parent abstract class.

**Comment by luizfbicalho** [comment]
I didn't get this update @JeanMeche  

I  updated all angular and dependencies and stil get 

 NG2007: Class is using Angular features but is not decorated. Please add an explicit Angular decorator.

19 export abstract class BaseFormlyComponent<T extends {}>

**Comment by JeanMeche** [comment]
You're on v16 ? Can you provide a small repro ?

**Comment by JeanMeche** [comment]
This is not exactly the same as the initial issue but you're right NG2007 is still thrown when using interfaces like `OnDestroy` on abstract classes.

**Comment by Harpush** [comment]
I think a distinction is required here:
1. Angular decorators like `@Input`
2. Angular lifecycle methods

In 1 it should fail without an angular decorator and it makes sense but in 2 there are cases where I am not going to use the class directly in DI but extend it and there is no harm doing it.
What can be added is maybe a check at usage site and not declaration site. If a constructor inject a class which isn't decorated or has no constructor and inherits an undecorated class it will fail to build. Thus eliminate all the false positives. I guess the exact nuance should be tweaked a bit as to how exactly to identify those cases.

**Comment by luizfbicalho** [comment]
I don't think that an abstract class should be in the DI


```ts
//@Injectable() if uncomment this results in error
export abstract class BaseFormlyComponent<T extends {}>
  implements OnInit
{
ngOnInit(): void {// some code here}
}

```
